### DISCUSSION TITLE
I like SSG, but I'd like to avoid loading states on client side navigation

### DISCUSSION THREAD

**Discussion by fabb** [vercel/next.js#12307] [Help] [answered]
SSG has reduced response times and required server resources drastically. I like it a lot. I have one issue though: client-side-navigation UX.

Imagine a non-SEO relevant user area, for example a "My Watch List" page. We'd statically render all the non-user-specific parts (general layout, header and such) and show a spinner for the central piece of the page, the watch list. This is fine when the user lands on this page via server-routing (which does not happen too often, maybe just by clicking links in emails). But when the user goes there via client-routing, it feels annoying to see the spinner, or an empty area that gets populated later. Imagine navigating between different pages in his/her user area, for example between "My Watch List" and "My Profile". Seeing a spinner/empty area  every time does not feel too nice of a UX.

Do you have any better solutions for this? From SSR pages, I'm used that a client-side navigation keeps showing the old page (+ NProgress for example) until `getInitialProps` of the target page is loaded, and only then directly render the target page. No empty areas, in-between flashes.

Incremental SSG won't solve the problem unfortunately, since we cannot/do not want to pre-render user area pages for all users (also the urls are the same for all users). https://github.com/zeit/next.js/discussions/11552

Will maybe in the future Transitions in React enable such patterns? (I guess it would need to be integrated with the Router by next.js first). https://reactjs.org/docs/concurrent-mode-patterns.html#transitions

Anything I could do today though before that lands?

**Comment by leerob** [comment, marked as answer]
If the data has already been loaded, then you should cache it to prevent showing a spinner again. Check out [SWR](https://swr.now.sh/). This is what Vercel uses for its own dashboard.

**Reply by fabb** [reply]
Does this really work with SSG pages? I fear that on client side navigation, I would first see the statically rendered html (including a spinner), and only after that SWR kicking in and showing the last loaded data. I hope to be wrong though, will try, thanks for the suggestion.

**Reply by fabb** [reply]
I was wrong, SWR + SSG works perfectly fine: https://github.com/zeit/swr/compare/master...fabb:experiments/ssg?expand=1

So it seems next.js is not using the static html when client side navigating, it would only use the static props from `getStaticProps`, but they can be empty when using SWR (in case the page is non-SEO-relevant).

The only issue with this approach is that when the data is not yet cached, it will show the spinner for the content. Would be nice to be able to avoid that.

Thanks, I'll try it in my project.

**Reply by fabb** [reply]
Now a way for SEO-relevant pages would be cool too, any ideas there? On these pages I do need to use `getStaticProps` and update data after hydration. On server side routing I guess there is nothing that I could do, but I imagine that on client side routing, I could rather use a cached value than the props from `getStaticProps`. Is there a way to initialize the SWR cache with data from `getStaticProps`?

**Reply by fabb** [reply]
Ah should be easy, just use the static data as a fallback:
```
const { data } = useSWR<{
    forks_count: number
    stargazers_count: number
    watchers: number
  }>('/api/data?id=' + id, fetch)
const dataOrFallback = data || fallbackDataFromGetStaticProps

return <MyComponent data={dataOrFallback} />
```

**Reply by fabb** [reply]
Ah, there is also `initialData`: https://github.com/zeit/swr/blob/master/README.md#ssr-with-nextjs

**Reply by fabb** [reply]
Very interesting, the vercel dashboard redirects to the login page directly via a script in the header to not have to wait for hydration and to prevent that not logged in users see the SSG html loading state at all:
```
		if (!document.cookie || document.cookie.indexOf('authorization=') === -1) {
              location.replace('/login?next=' + encodeURIComponent(location.pathname + location.search))
            } else {
              document.documentElement.classList.add('render')
            }
```

**Comment by fabb** [comment]
Unfortunately SWR does not help in cases where I cannot preload data for the next page. Maybe React Concurrent Mode will be a solution for this, using deferred values could help to only show the target page after some timeout where it hopefully already finished loading and rendering: https://reactjs.org/docs/concurrent-mode-patterns.html#deferring-a-value
